lib/Conch/Controller/DeviceReport.pm

```diff
@@ -19,54 +19,101 @@ with 'Conch::Role::MojoLog';
 
 =head2 process
 
-Processes the device report using the Legacy report code base
+Processes the device report, turning it into the various device_ tables as well
+as running validations
 
 =cut
 
 sub process ($c) {
-	my $device_report = $c->validate_input('DeviceReport');
-	return if not $device_report;
 	my $raw_report = $c->req->body;
+	my $schema = $c->schema;
 
-	my $maybe_hw;
+	my $unserialized_report = $c->validate_input('DeviceReport');
+	if(not $unserialized_report) {
+		return; # JSON validator will throw the proper error response
+	}
 
-	if ( $device_report->{device_type}
-		&& $device_report->{device_type} eq "switch" )
-	{
-		$maybe_hw = Conch::Model::HardwareProduct->lookup_by_name(
-			$device_report->{product_name}
-		);
-		return $c->status(409, {
-			error => "Hardware product name '".$device_report->{product_name}."' does not exist"
-		}) unless ($maybe_hw);
+	# Make sure the API and device report agree on who we're talking about
+	if ($c->param('id') ne $unserialized_report->{serial_number}) {
+		return $c->render(status => 422, json => {
+			error => "Serial number provided to the API does not match the report data."
+		});
+	}
 
+	my $hw;
+	# Make sure that the remote side is telling us about a hardware product we understand
+	if ($unserialized_report->{device_type} && $unserialized_report->{device_type} eq "switch") {
+		$hw = $schema->resultset('HardwareProduct')->find({
+			name => $unserialized_report->{product_name}
+		});
 	} else {
-		$maybe_hw = Conch::Model::HardwareProduct->lookup_by_sku(
-			$device_report->{sku}
-		);
+		$hw = $schema->resultset('HardwareProduct')->find({
+			sku => $unserialized_report->{sku}
+		});
+
+		if(not $hw) {
+		  	$hw = $schema->resultset('HardwareProduct')->find({
+				legacy_product_name => $unserialized_report->{product_name}
+			});
+		}
+	}
 
-		return $c->status(409, {
-			error => "Hardware product SKU '".$device_report->{sku}."' does not exist"
-		}) unless ($maybe_hw);
+	if(not $hw) {
+		return $c->render(status => 409, json => {
+			error => "Could not locate hardware product"
+		});
+	}
 
+	if(not $hw->hardware_product_profile) {
+		return $c->render(status => 409, json => {
+			error => "Hardware product does not contain a profile"
+		});
 	}
 
-	# Use the old device report recording and device validation code for now.
-	# This will be removed when OPS-RFD 22 is implemented
-	$c->log->debug("Recording device report");
-	my ( $device, $report_id ) = $c->_record_device_report(
-		$device_report,
-		$raw_report
+	my $existing_device = $schema->resultset('Device')->find({
```

Comment:
This is already available in `$c->stash('current_device')` (and in the act of looking that up, it also verified that the user has permission to see this device, by virtue of their workspace(s) and/or previous relay reports -- although perhaps in this case that second part of the verification is not wanted?)

Reply:
No it's not. `POST /device/:id` is not part of that `under`.

Reply:
The patch to `Route/Device.pm` included in this PR actually moves that route above the `under` declaration to make it more clear that it stands alone. 

Current behavior requires this route to get called even if the device doesn't exist.

Reply:
k

Reply:
should we still be verifying that the device, if it already exists, is under a workspace the user has permission to access?

Reply:
That's a question for a different PR. The code prior to this PR did not make that sort of permissions check so neither does the code in the PR.